apex/contrib/optimizers/distributed_fused_adam.py

```diff
@@ -862,6 +1150,64 @@ def _grad_copy(self, param):
         # Free param grad buffer
         param.grad = None
 
+    def _param_copy(self, params):
+        """Update parameters with values from parameter buckets"""
+
+        # Get parameter fragments to be synchronized
+        if isinstance(params, torch.Tensor):
+            params = [params]
+        fragments = []
+        for param in params:
+            if 'fragments' in self.state[param]:
+                fragments.extend(
+                    fragment
+                    for fragment in self.state[param]['fragments']
+                    if fragment.bucket_id in self._params_buckets
+                )
+
+        # Make sure all needed buckets have been synchronized
+        buckets = collections.OrderedDict()
```

Comment:
fwiw recent Python keeps the insertion order as per https://mail.python.org/pipermail/python-dev/2017-December/151283.html

Reply:
Huh, that's pretty neat. I'm a little scared to rely on it though.